jetty-server/src/main/java/org/eclipse/jetty/server/ProxyConnectionFactory.java

```diff
@@ -46,252 +45,281 @@
-        private int _length;
 
-        protected ProxyProtocolV1Connection(EndPoint endp, Connector connector, String next, ByteBuffer buffer)
-        {
-            super(endp, connector.getExecutor());
-            _connector = connector;
-            _next = next;
-            _length = buffer.remaining();
-            parse(buffer);
+            if (LOG.isDebugEnabled())
+                LOG.debug("Proxy v1 detection succeeded");
+            return Detection.RECOGNIZED;
         }
 
         @Override
-        public void onOpen()
+        public Connection newConnection(Connector connector, EndPoint endp)
         {
-            super.onOpen();
-            fillInterested();
+            ConnectionFactory nextConnectionFactory = findNextConnectionFactory(_nextProtocol, connector, getProtocol(), endp);
+            return configure(new ProxyProtocolV1Connection(endp, connector, nextConnectionFactory), connector, endp);
         }
 
-        private boolean parse(ByteBuffer buffer)
+        private static class ProxyProtocolV1Connection extends AbstractConnection implements Connection.UpgradeFrom, Connection.UpgradeTo
         {
-            // Parse fields
-            while (buffer.hasRemaining())
+            // 0     1 2       3       4 5 6
+            // 98765432109876543210987654321
+            // PROXY P R.R.R.R L.L.L.L R Lrn
+
+            private final Connector _connector;
+            private final ConnectionFactory _next;
+            private final ByteBuffer _buffer;
+            private final StringBuilder _builder = new StringBuilder();
+            private final String[] _fields = new String[6];
+            private int _index;
+            private int _length;
+
+            private ProxyProtocolV1Connection(EndPoint endp, Connector connector, ConnectionFactory next)
             {
-                byte b = buffer.get();
-                if (_index < 6)
+                super(endp, connector.getExecutor());
+                _connector = connector;
+                _next = next;
+                _buffer = _connector.getByteBufferPool().acquire(getInputBufferSize(), true);
+            }
+
+            @Override
+            public void onFillable()
+            {
+                if (LOG.isDebugEnabled())
+                    LOG.debug("Proxy v1 onFillable current index = ", _index);
+                try
                 {
-                    if (b == ' ' || b == '\r')
-                    {
-                        _fields[_index++] = _builder.toString();
-                        _builder.setLength(0);
-                        if (b == '\r')
-                            _index = 6;
-                    }
-                    else if (b < ' ')
-                    {
-                        LOG.warn("Bad character {} for {}", b & 0xFF, getEndPoint());
-                        close();
-                        return false;
-                    }
-                    else
+                    while (_index < 7)
```

Comment:
maybe avoid hard coded 7.   Use a descriptive constant or is this `<=_fields.length`?

Reply:
7 is the field index of the line feed character. It used to be hardcoded, but I can make that a constant.